File: HW/HW7/HW7.java
import java.util.*;


public class HW7 {
	static class compareStartTimes implements Comparator<int[]>{
		public int compare(int[] a, int[] b) {
			return a[0] - b[0];
		}
	}
	
	public static int minMeetingRooms(int[][] intervals) {
        if (intervals == null || intervals.length == 0) {
            return 0;
        }

        Arrays.sort(intervals, new compareStartTimes());

        PriorityQueue<Integer> minHeap = new PriorityQueue<>();

        for (int[] interval : intervals) {
            int startTime = interval[0];

            if (!minHeap.isEmpty() && minHeap.peek() <= startTime) {
                minHeap.poll();
            }

            minHeap.offer(interval[1]);
        }
        return minHeap.size();
    }

    public static void main(String[] args) {

        int[][] intervals1 = {{0, 30}, {5, 10}, {15, 20}};
        System.out.println(minMeetingRooms(intervals1)); 

        int[][] intervals2 = {{0, 1}, {1, 2}, {2, 3}};
        System.out.println(minMeetingRooms(intervals2));
    }
}

File: HW/HW7/HW7.py
import heapq
from typing import List

class MeetingRooms:
    def get_start_time(self, interval):
        return interval[0]

    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        intervals.sort(key=self.get_start_time)
        min_heap = []

        for interval in intervals:
            start_time = interval[0]

            if min_heap and min_heap[0] <= start_time:
                heapq.heappop(min_heap)

            heapq.heappush(min_heap, interval[1])

        return len(min_heap)

# Example usage
if __name__ == "__main__":
    solution = MeetingRooms()

    intervals1 = [[0, 30], [5, 10], [15, 20]]
    print(solution.minMeetingRooms(intervals1))  # Output: 2

    intervals2 = [[0, 1], [1, 2], [2, 3]]
    print(solution.minMeetingRooms(intervals2))  # Output: 1
